### PR TITLE
[GEOT-7428] Support of jsonPointer function for Oracle

### PR DESCRIPTION
**Pull request by Mike7311**
[GEOT-7428](https://osgeo-org.atlassian.net/browse/GEOT-7428) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

Added delegation of jsonPointer function unto Oracle’s JSON_VALUE function

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Review by mprins — APPROVED** [review]
lgtm

File: modules/library/main/src/main/java/org/geotools/filter/visitor/JsonPointerFilterSplittingVisitor.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            DuplicatingFilterVisitor duplicating = new DuplicatingFilterVisitor();
+            Function duplicated = (Function) expression.accept(duplicating, null);
+            // if constant can encode
+            Object result = param.evaluate(null);
+            FilterFactory2 ff = CommonFactoryFinder.getFilterFactory2();
+            // setting constant expression evaluated to literal
+            duplicated.getParameters().set(paramIdx, ff.literal(result));
+            return duplicated;
+        }
+        return expression;
```
Comment (aaime) [inline modules/library/main/src/main/java/org/geotools/filter/visitor/JsonPointerFilterSplittingVisitor.java:74]: The code looks inconsistent to me. ``constantParameterToLiteral`` is more general, checks if the first parameter is equivalent to a static expression, e.g., it's able to recognize that a ``env`` function is equivalent to a constant for the current request. ``supports`` instead is checking for a straight literal, would claim ``env`` is not supported and thus move the JSON function to the post filter, right?

File: modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleFilterToSQL.java [hunk 2]
```diff
@@ -235,11 +243,42 @@ public Object visit(Function function, Object extraData) {
...
+            }
         }
 
         return super.visit(function, extraData);
     }
 
+    private String encodeJsonPointer(Function jsonPointer, Object extraData) {
+        Expression json = getParameter(jsonPointer, 0, true);
+        Expression pointer = getParameter(jsonPointer, 1, true);
+        if (json instanceof PropertyName && pointer instanceof Literal) {
```
Comment (aaime) [inline modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleFilterToSQL.java:261]: Here is also assumes it has to be a straight literal. Ah, another important thing, is when FilterSimplifier is applied in the JDBC code. If it's applied before the splitting, then you'll receive a Literal anyways, as the ``env(...)`` has already been replaced by a Literal in the simplifier.

File: modules/library/main/src/main/java/org/geotools/filter/visitor/JsonPointerFilterSplittingVisitor.java (hunk 1 above)
Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/filter/visitor/JsonPointerFilterSplittingVisitor.java:74]: Also, I suggest adding a test for the splitter, with supported and non-supported cases.

**Review by aaime — CHANGES_REQUESTED** [review]
The PR needs to be rebased on top of main.
More feedback below.

File: modules/library/main/src/main/java/org/geotools/filter/visitor/JsonPointerFilterSplittingVisitor.java (hunk 1 above)
Comment (Mike7311) [reply modules/library/main/src/main/java/org/geotools/filter/visitor/JsonPointerFilterSplittingVisitor.java]: > The code looks inconsistent to me. `constantParameterToLiteral` is more general, checks if the first parameter is equivalent to a static expression, e.g., it's able to recognize that a `env` function is equivalent to a constant for the current request. `supports` instead is checking for a straight literal, would claim `env` is not supported and thus move the JSON function to the post filter, right?

constantParameterToLiteral is called from the if block, which is checking the expression to be JsonPointerFunction in [JsonPointerFilterSplittingVisitor#visit](https://github.com/Mike7311/geotools/blob/7c9eb32894310b8322fb8e425480d7e6bcc64187/modules/library/main/src/main/java/org/geotools/filter/visitor/JsonPointerFilterSplittingVisitor.java#L40). So env will not pass this check.

I will add tests for supported/non-supported cases

Comment (Mike7311) [reply modules/library/main/src/main/java/org/geotools/filter/visitor/JsonPointerFilterSplittingVisitor.java]: Tests added

File: modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleFilterToSQL.java (hunk 2 above)
Comment (Mike7311) [reply modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleFilterToSQL.java:261]: I think the reply to the previous comment also applies here.

**Pull request merged**